File: libraries/python/src/pandaprobe/tracing/context.py
"""TraceContext — manages the lifecycle of a single trace."""

from __future__ import annotations

import logging
from contextvars import ContextVar, Token
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any
from uuid import uuid4

from pandaprobe.schemas import SpanKind, TraceData, TraceStatus
from pandaprobe.tracing.session import get_current_session_id, get_current_user_id
from pandaprobe.validation import warn_if_invalid_messages

if TYPE_CHECKING:
    from pandaprobe.client import Client
    from pandaprobe.tracing.span import SpanContext

logger = logging.getLogger("pandaprobe")

# Global contextvar tracking the active trace context.
_current_trace: ContextVar[TraceContext | None] = ContextVar("_current_trace", default=None)
# Global contextvar tracking the stack of active span IDs.
_span_stack: ContextVar[list[str]] = ContextVar("_span_stack", default=[])


def get_current_trace() -> TraceContext | None:
    return _current_trace.get(None)


def get_span_stack() -> list[str]:
    return _span_stack.get()


class TraceContext:
    """Context manager that wraps a trace lifecycle.

    Input must be ``{"messages": [{"role": "user", "content": "..."}]}``
    containing only the current turn's user message.
    Output must be ``{"messages": [{"role": "assistant", "content": "..."}]}``.

    Usage::

        with client.trace(
            "my-trace",
            input={"messages": [{"role": "user", "content": "hello"}]},
        ) as t:
            with t.span("llm-call", kind="LLM") as s:
                s.set_input({"messages": [
                    {"role": "system", "content": "You are helpful."},
                    {"role": "user", "content": "hello"},
                ]})
                ...
                s.set_output({"messages": [{"role": "assistant", "content": "hi!"}]})
            t.set_output({"messages": [{"role": "assistant", "content": "hi!"}]})
    """

    def __init__(
        self,
        client: Client,
        name: str,
        *,
        input: Any = None,
        session_id: str | None = None,
        user_id: str | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> None:
        self._client = client
        self._trace_id = str(uuid4())
        self._name = name
        self._input = input
        warn_if_invalid_messages(input, "trace input")
        self._output: Any = None
        self._session_id = session_id if session_id is not None else get_current_session_id()
        self._user_id = user_id if user_id is not None else get_current_user_id()
        self._tags = tags or []
        self._metadata = metadata or {}
        self._status: TraceStatus = TraceStatus.COMPLETED
        self._error: str | None = None
        self._started_at: datetime | None = None
        self._ended_at: datetime | None = None
        self._spans: list = []
        self._token: Token[TraceContext | None] | None = None
        self._stack_token: Token[list[str]] | None = None

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def trace_id(self) -> str:
        return self._trace_id

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    def __enter__(self) -> TraceContext:
        self._started_at = datetime.now(timezone.utc)
        self._token = _current_trace.set(self)
        self._stack_token = _span_stack.set([])
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:  # noqa: ANN001
        self._ended_at = datetime.now(timezone.utc)
        if exc_type is not None:
            self._status = TraceStatus.ERROR
            self._error = str(exc_val)

        self._finalize()

        if self._token is not None:
            _current_trace.reset(self._token)
        if self._stack_token is not None:
            _span_stack.reset(self._stack_token)

        return None  # don't suppress exceptions

    async def __aenter__(self) -> TraceContext:
        return self.__enter__()

    async def __aexit__(self, exc_type, exc_val, exc_tb):  # noqa: ANN001
        return self.__exit__(exc_type, exc_val, exc_tb)

    # ------------------------------------------------------------------
    # Span factory
    # ------------------------------------------------------------------

    def span(
        self,
        name: str,
        *,
        kind: str | SpanKind = SpanKind.OTHER,
        model: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> SpanContext:
        from pandaprobe.tracing.span import SpanContext

        return SpanContext(
            trace_ctx=self,
            name=name,
            kind=SpanKind(kind) if isinstance(kind, str) else kind,
            model=model,
            metadata=metadata,
        )

    # ------------------------------------------------------------------
    # Setters
    # ------------------------------------------------------------------

    def set_output(self, output: Any) -> None:
        warn_if_invalid_messages(output, "trace output")
        self._output = output

    def set_input(self, input: Any) -> None:
        warn_if_invalid_messages(input, "trace input")
        self._input = input

    def set_metadata(self, metadata: dict[str, Any]) -> None:
        self._metadata.update(metadata)

    def set_status(self, status: TraceStatus | str) -> None:
        self._status = TraceStatus(status)

    def add_span_data(self, span_data) -> None:  # noqa: ANN001
        self._spans.append(span_data)

    # ------------------------------------------------------------------
    # Finalization
    # ------------------------------------------------------------------

    def _finalize(self) -> None:
        """Build TraceData and enqueue for background submission."""
        try:
            trace = TraceData(
                trace_id=self._trace_id,
                name=self._name,
                status=self._status,
                input=self._input,
                output=self._output,
                metadata=self._metadata,
                started_at=self._started_at,
                ended_at=self._ended_at,
                session_id=self._session_id,
                user_id=self._user_id,
                tags=self._tags,
                spans=self._spans,
            )
            self._client.log_trace(trace)
        except Exception as exc:
            logger.error("PandaProbe failed to finalize trace: %s", exc)

File: libraries/typescript/src/tracing/context.ts
/**
 * TraceContext — manages the lifecycle of a single trace.
 *
 * Input must be `{messages: [{role: "user", content: "..."}]}` containing only
 * the current turn's user message. Output must be
 * `{messages: [{role: "assistant", content: "..."}]}`.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import { logger } from "../logger.js";
import { type SpanData, SpanKind, TraceStatus as TS, TraceData, type TraceStatus, newUuid } from "../schemas.js";
import { isThenable } from "../util.js";
import { extractLastAssistantMessage, warnIfInvalidMessages } from "../validation.js";
import { getCurrentSessionId, getCurrentUserId } from "./session.js";
import { SpanContext, type SpanContextOptions } from "./span.js";

/** What a TraceContext needs from the client (avoids a circular import). */
export interface TraceClient {
  logTrace(trace: TraceData): void;
}

interface TraceStore {
  trace: TraceContext | null;
}

const traceStorage = new AsyncLocalStorage<TraceStore>();

export function getCurrentTrace(): TraceContext | null {
  return traceStorage.getStore()?.trace ?? null;
}

/** The span stack of the active trace (each trace owns its own stack). */
export function getSpanStack(): string[] {
  return getCurrentTrace()?.spanStack ?? [];
}

export interface TraceContextOptions {
  input?: unknown;
  sessionId?: string | null;
  userId?: string | null;
  tags?: string[];
  metadata?: Record<string, unknown>;
}

export class TraceContext {
  private readonly client: TraceClient;
  private readonly _traceId: string;
  private readonly name: string;
  private input: unknown;
  private output: unknown = null;
  private readonly sessionId: string | null;
  private readonly userId: string | null;
  private readonly tags: string[];
  private readonly metadata: Record<string, unknown>;
  private status: TraceStatus = TS.COMPLETED;
  private error: string | null = null;
  private startedAt: Date | null = null;
  private endedAt: Date | null = null;
  private readonly spans: SpanData[] = [];
  readonly spanStack: string[] = [];
  private prevStore: TraceStore | undefined;
  private entered = false;
  private isEnded = false;

  constructor(client: TraceClient, name: string, options: TraceContextOptions = {}) {
    this.client = client;
    this._traceId = newUuid();
    this.name = name;
    this.input = options.input ?? null;
    warnIfInvalidMessages(this.input, "trace input");
    this.sessionId = options.sessionId ?? getCurrentSessionId();
    this.userId = options.userId ?? getCurrentUserId();
    this.tags = options.tags ?? [];
    this.metadata = options.metadata ?? {};
  }

  get traceId(): string {
    return this._traceId;
  }

  // ------------------------------------------------------------------
  // Lifecycle — imperative (start/end) and callback (run)
  // ------------------------------------------------------------------

  /**
   * Enter the trace context imperatively. Pair with `end()`.
   *
   * When *enterStore* is true (the default, for the imperative `startTrace`
   * API) the trace is pushed onto the AsyncLocalStorage store so ambient
   * lookups (`getCurrentTrace`) find it. Wrappers that create a standalone
   * trace pass `false` — their spans nest via this trace's own `spanStack`, so
   * they must NOT mutate the global store (`enterWith` would leak the trace
   * into the caller's async frame and contaminate later, unrelated calls).
   */
  start(enterStore = true): this {
    this.startedAt = new Date();
    if (enterStore) {
      this.prevStore = traceStorage.getStore();
      this.entered = true;
      traceStorage.enterWith({ trace: this });
    }
    return this;
  }

  /**
   * End the trace, finalize, and restore the previous ambient context.
   *
   * Imperative `start()`/`end()` is intended for LIFO (properly nested) usage,
   * the analog of Python's `with`. For overlapping or concurrent traces prefer
   * the scoped {@link run} / `withTrace` form. To stay safe under out-of-order
   * end() the restore is guarded: it only rewrites the ambient store when THIS
   * trace is still the active one (so ending an inner trace out of order can't
   * wipe a newer active trace), and it never restores to an already-ended
   * parent (which would resurrect a finished trace as "current").
   */
  end(error?: unknown): void {
    this.endedAt = new Date();
    if (error !== undefined && error !== null) {
      this.status = TS.ERROR;
      this.error = String(error instanceof Error ? error.message : error);
    }
    this.isEnded = true;
    this.finalize();
    if (this.entered && getCurrentTrace() === this) {
      const parent = this.prevStore?.trace ?? null;
      traceStorage.enterWith({ trace: parent?.isEnded ? null : parent });
    }
  }

  /** Run *fn* within this trace's context (callback form). */
  async run<T>(fn: (ctx: TraceContext) => T | Promise<T>): Promise<T> {
    return traceStorage.run({ trace: this }, async () => {
      this.startedAt = new Date();
      try {
        return await fn(this);
      } catch (exc) {
        this.status = TS.ERROR;
        this.error = String(exc instanceof Error ? exc.message : exc);
        throw exc;
      } finally {
        this.endedAt = new Date();
        this.finalize();
      }
    });
  }

  /**
   * Wrap a sync-or-async invocation in this trace's context (decorator form).
   *
   * Uses `AsyncLocalStorage.run` so async continuations inherit the trace store
   * without leaking it to the caller. Preserves the original return type:
   * a sync invocation returns synchronously, an async one returns a Promise.
   */
  instrument<T>(invoke: () => T, captureOutput: boolean): T {
    return traceStorage.run({ trace: this }, (): T => {
      this.startedAt = new Date();
      let result: T;
      try {
        result = invoke();
      } catch (exc) {
        this.failAndFinalize(exc);
        throw exc;
      }
      if (isThenable(result)) {
        const wrapped = (result as unknown as Promise<unknown>).then(
          (value) => {
            if (captureOutput) {
              this.setOutput(extractLastAssistantMessage(value));
            }
            this.endedAt = new Date();
            this.finalize();
            return value;
          },
          (exc) => {
            this.failAndFinalize(exc);
            throw exc;
          },
        );
        return wrapped as T;
      }
      if (captureOutput) {
        this.setOutput(extractLastAssistantMessage(result));
      }
      this.endedAt = new Date();
      this.finalize();
      return result;
    });
  }

  private failAndFinalize(exc: unknown): void {
    this.status = TS.ERROR;
    this.error = String(exc instanceof Error ? exc.message : exc);
    this.endedAt = new Date();
    this.finalize();
  }

  // ------------------------------------------------------------------
  // Span factory
  // ------------------------------------------------------------------

  span(name: string, options: SpanContextOptions = {}): SpanContext {
    return new SpanContext(this, name, {
      kind: options.kind ?? SpanKind.OTHER,
      model: options.model,
      metadata: options.metadata,
    });
  }

  // ------------------------------------------------------------------
  // Setters
  // ------------------------------------------------------------------

  setOutput(output: unknown): void {
    warnIfInvalidMessages(output, "trace output");
    this.output = output;
  }

  setInput(input: unknown): void {
    warnIfInvalidMessages(input, "trace input");
    this.input = input;
  }

  setMetadata(metadata: Record<string, unknown>): void {
    Object.assign(this.metadata, metadata);
  }

  setStatus(status: TraceStatus): void {
    this.status = status;
  }

  addSpanData(spanData: SpanData): void {
    this.spans.push(spanData);
  }

  // ------------------------------------------------------------------
  // Finalization
  // ------------------------------------------------------------------

  private finalize(): void {
    try {
      const trace = new TraceData({
        traceId: this._traceId,
        name: this.name,
        status: this.status,
        input: this.input,
        output: this.output,
        metadata: this.metadata,
        startedAt: this.startedAt ?? new Date(),
        endedAt: this.endedAt,
        sessionId: this.sessionId,
        userId: this.userId,
        tags: this.tags,
        spans: this.spans,
      });
      this.client.logTrace(trace);
    } catch (exc) {
      logger.error(`failed to finalize trace: ${String(exc)}`);
    }
  }
}
